[UnifiedPDF] Add ability to start focusing next/previous text annotations.

sammygill opened the pull request:
#### 3b5f85622e9c64d8e6e1c748eeff3b9bdfafa82f
<pre>
[UnifiedPDF] Add ability to start focusing next/previous text annotations.
<a href="https://bugs.webkit.org/show_bug.cgi?id=268349">https://bugs.webkit.org/show_bug.cgi?id=268349</a>
<a href="https://rdar.apple.com/problem/121890426">rdar://problem/121890426</a>

Reviewed by Simon Fraser.

When a text annotation is currently active, the user should be able
to focus the next/previous one using the appropriate keyboard shortcut.
Effectivley, what we will need to do is iterate over the annotations in
the appropriate direction until we find a text annotation that can be
interacted with (this could be the same annotation if there are no
others to be found).

We start by checking the current page the active annotation is on. If
there are any remaining annotations on the page, then we will iterate
over them to see if the next one is on the page. If we could not find an
appropriate text annotation on the page then we repeat the process by
going to the next page in the appropriate direction. This process is
repeated until we find an appropriate text annotation to return. There
should always be at least one text annotation to return: the original
one.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::findFirstTextAnnotationStartingAtIndex):
(WebKit::UnifiedPDFPlugin::nextTextAnnotation const):
(WebKit::UnifiedPDFPlugin::focusNextAnnotation):
(WebKit::UnifiedPDFPlugin::focusPreviousAnnotation):

Canonical link: <a href="https://commits.webkit.org/274337@main">https://commits.webkit.org/274337@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e1bb27ecb6f0e3be5befe24b1498de42377bd79

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41265 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20477 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15011 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32506 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39307 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14879 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12919 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42541 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35177 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38725 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13544 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11198 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36933 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8684 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/14678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -42,6 +42,10 @@ struct PDFContextMenu;
 class WebFrame;
 class WebMouseEvent;
 
+enum class AnnotationSearchDirection : uint8_t {
```
Can be packed more tightly
```suggestion
enum class AnnotationSearchDirection : bool {
```

sammygill replied:
Done!

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -237,6 +242,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     bool m_inMagnificationGesture { false };
 };
 
+RetainPtr<PDFAnnotation> getFirstTextAnnotationStartingAtIndex(const RetainPtr<NSArray>&, unsigned startingIndex, AnnotationSearchDirection);
```
The first argument should get a name since it's not clear what it is

aprotyas replied:
Well, it is clear to me now, but it won't be in the future 😛 And it's a generic enough type to warrant an identifier

sammygill replied:
Oops I totally agree I think this was a bit of an oversight

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
     return pageSpaceBounds;
 }
 
+RetainPtr<PDFAnnotation> getFirstTextAnnotationStartingAtIndex(const RetainPtr<NSArray>& annotations, unsigned startingIndex, AnnotationSearchDirection searchDirection)
```
This could just be a `static` function in this translation unit, right?

sammygill replied:
Yes! Done!

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+        return [NSIndexSet indexSetWithIndexesInRange: NSMakeRange(0, startingIndex + 1)];
+    }();
+
+    auto searchResult = [annotations indexOfObjectAtIndexes: indexRange options: searchDirection == AnnotationSearchDirection::Forward ? 0 : NSEnumerationReverse passingTest:^BOOL(PDFAnnotation* annotation, NSUInteger index, BOOL *stop) {
```
Nit: since `index` and `stop` aren't used.
```suggestion
    auto searchResult = [annotations indexOfObjectAtIndexes: indexRange options: searchDirection == AnnotationSearchDirection::Forward ? 0 : NSEnumerationReverse passingTest:^BOOL(PDFAnnotation* annotation, NSUInteger, BOOL *) {
```

sammygill replied:
Changed

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    if (!indexForCurrentPage)
+        return nullptr;
+
+    RetainPtr<PDFAnnotation> nextAnnotation = nullptr;
```
Nit: this default initializes to `nullptr`, so `RetainPtr<...> value;` is fine

sammygill replied:
Done

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+        nextPageToSearchIndex = getNextPageToSearchIndex(nextPageToSearchIndex);
+        RetainPtr<PDFPage> nextPage = m_documentLayout.pageAtIndex(nextPageToSearchIndex);
+        if (!nextPage)
+            continue;
```
Shouldn't we bail out here?

sammygill replied:
Maybe.... I changed it but not really sure what the "correct," decision is in this particular scenario

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    if (!previousTextAnnotation || previousTextAnnotation == m_activeAnnotation->annotation())
+        return;
+    setActiveAnnotation(previousTextAnnotation.get());
 }
```
Should we be focusing on the first / last annotation (respectively) if there is no active annotation?

sammygill replied:
I don't think so. The current behavior seems to be that the focusing keyboard shortcuts do not do anything if an annotation is currently not active

aprotyas reviewed (commented):
Looks good! I have some feedback. Also, some of the objC calls have inconsistent spacing!

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -63,6 +67,7 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     void setActiveAnnotation(RetainPtr<PDFAnnotation>&&) final;
     void focusNextAnnotation() final;
     void focusPreviousAnnotation() final;
+    RetainPtr<PDFAnnotation> getNextTextAnnotation(AnnotationSearchDirection) const;
```
Getters in WebKit for not have a "get" prefix, per coding style.

sammygill replied:
Fixed

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+RetainPtr<PDFAnnotation> UnifiedPDFPlugin::getNextTextAnnotation(AnnotationSearchDirection searchDirection) const
+{
+    ASSERT(m_activeAnnotation);
+    RetainPtr<PDFAnnotation> currentAnnotation = m_activeAnnotation->annotation();
```
I don't believe you need `<PDFAnnotation>` with C++17. Samec comment applies below for other smart pointers.

sammygill replied:
Fixed

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    if (indexOfCurrentAnnotation == NSNotFound)
+        return nullptr;
+
+    auto isForwardSearchDirection = searchDirection == AnnotationSearchDirection::Forward;
```
`bool` is as long as `auto` and clearer IMO.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    RetainPtr<PDFAnnotation> previousTextAnnotation = getNextTextAnnotation(AnnotationSearchDirection::Backward);
+    if (!previousTextAnnotation || previousTextAnnotation == m_activeAnnotation->annotation())
+        return;
+    setActiveAnnotation(previousTextAnnotation.get());
```
`WTFMove(previousTextAnnotation)` to avoid refcounting churn.

sammygill replied:
Fixed

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    RetainPtr<PDFAnnotation> nextTextAnnotation = getNextTextAnnotation(AnnotationSearchDirection::Forward);
+    if (!nextTextAnnotation || nextTextAnnotation == m_activeAnnotation->annotation())
+        return;
+    setActiveAnnotation(nextTextAnnotation.get());
```
`WTFMove()`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
     return pageSpaceBounds;
 }
 
+static RetainPtr<PDFAnnotation> findFirstTextAnnotationStartingAtIndex(const RetainPtr<NSArray>& annotations, unsigned startingIndex, AnnotationSearchDirection searchDirection)
```
There's a system setting to focus controls that are not just text fields; we should respect that, and focus things like checkboxes when it's enabled.

sammygill replied:
This current implementation matches shipping behavior with that setting toggled on and off. I created a bugzilla to properly support focus rings with this setting on: https://bugs.webkit.org/show_bug.cgi?id=269004

sammygill committed:
[UnifiedPDF] Add ability to start focusing next/previous text annotations.
https://bugs.webkit.org/show_bug.cgi?id=268349
rdar://problem/121890426

Reviewed by Simon Fraser.

When a text annotation is currently active, the user should be able
to focus the next/previous one using the appropriate keyboard shortcut.
Effectivley, what we will need to do is iterate over the annotations in
the appropriate direction until we find a text annotation that can be
interacted with (this could be the same annotation if there are no
others to be found).

We start by checking the current page the active annotation is on. If
there are any remaining annotations on the page, then we will iterate
over them to see if the next one is on the page. If we could not find an
appropriate text annotation on the page then we repeat the process by
going to the next page in the appropriate direction. This process is
repeated until we find an appropriate text annotation to return. There
should always be at least one text annotation to return: the original
one.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::findFirstTextAnnotationStartingAtIndex):
(WebKit::UnifiedPDFPlugin::nextTextAnnotation const):
(WebKit::UnifiedPDFPlugin::focusNextAnnotation):
(WebKit::UnifiedPDFPlugin::focusPreviousAnnotation):

Canonical link: https://commits.webkit.org/274337@main
